python/rascaline/splines.py

```diff
@@ -0,0 +1,254 @@
+# -*- coding: utf-8 -*-
+from collections import namedtuple
+
+import numpy as np
+
+
+def generate_splines(
+    radial_basis,
+    radial_basis_derivatives,
+    max_radial,
+    max_angular,
+    cutoff_radius,
+    n_spline_points=None,
+    requested_accuracy=1e-6,
+):
+
+    """Spline generator for tabulated radial integrals.
+
+    Besides some self-explanatory parameters, this function takes as inputs two
+    functions, namely radial_basis and radial_basis_derivatives. These must be
+    able to calculate the radial basis functions by taking n, l, and r as their
+    inputs, where n and l are integers and r is a numpy 1-D array that contains
+    the spline points at which the radial basis function (or its derivative)
+    needs to be evaluated. These functions should return a numpy 1-D array
+    containing the values of the radial basis function (or its derivative)
+    corresponding to the specified n and l, and evaluated at all points in the
+    r array. If specified, n_spline_points determines how many spline points
+    will be used for each splined radial basis function. Alternatively, the user
+    can specify a requested accuracy. Spline points will be added until either
+    the relative error or the absolute error fall below the requested accuracy on
+    average across all radial basis functions.
+    """
+
+    def value_evaluator_3D(positions):
+        spline_f = []
+        for el in range(max_angular + 1):
+            for n in range(max_radial):
+                spline_f_single = radial_basis(n, el, positions)
+                spline_f.append(spline_f_single)
+        spline_f = np.array(spline_f).T
+        spline_f = spline_f.reshape(len(positions), max_angular + 1, max_radial)
+        return spline_f
+
+    def derivative_evaluator_3D(positions):
+        spline_df = []
+        for el in range(max_angular + 1):
+            for n in range(max_radial):
+                spline_df_single = radial_basis_derivatives(n, el, positions)
+                spline_df.append(spline_df_single)
+        spline_df = np.array(spline_df).T
+        spline_df = spline_df.reshape(len(positions), max_angular + 1, max_radial)
+        return spline_df
+
+    if n_spline_points is not None:  # if user specifies the number of spline points
+        spline_x = np.linspace(0.0, cutoff_radius, n_spline_points)  # spline positions
+        spline_f = value_evaluator_3D(spline_x)
+        spline_df = derivative_evaluator_3D(spline_x)
+    else:
+        dynamic_spliner = DynamicSpliner(
+            0.0,
+            cutoff_radius,
+            value_evaluator_3D,
+            derivative_evaluator_3D,
+            requested_accuracy,
+        )
+        spline_x, spline_f, spline_df = dynamic_spliner.spline()
+
+    # Due to the inner workings of rascaline,
+    # values and derivatives need to be multiplied by 1/(4pi):
+    spline_f = spline_f / (4.0 * np.pi)
+    spline_df = spline_df / (4.0 * np.pi)
```

Comment:
Ok, we need to resolve this before merging this PR. I'm fine with putting the 4pi factor either inside or outside of the radial integral definition, but we need to document what's expected as a radial integral. @kvhuguenin do you have a mathematical expression we can put in the documentation?